### ISSUE TITLE
Phase 39.4 — AttentionVisualizer: Attention Rollout, GradCAM, LRP & Multi-Head Visualization

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#791]
## Phase 39.4 — AttentionVisualizer

> Attention Rollout, GradCAM, Layer-wise Relevance Propagation, Multi-Head Visualization

### Overview

The `AttentionVisualizer` makes transformer and attention-based model internals transparent through multiple visualization techniques. Implements Abnar & Zuidema (2020) attention rollout, Selvaraju et al. (2017) GradCAM adapted for transformers, Bach et al. (2015) Layer-wise Relevance Propagation (LRP), and BertViz-style multi-head attention visualization.

### Architecture

```
┌──────────────────────────────────────────────┐
│           AttentionVisualizer                │
│  ┌──────────┐  ┌──────────┐  ┌────────────┐ │
│  │Attention │  │ GradCAM  │  │   LRP      │ │
│  │ Rollout  │  │Transformer│  │  Engine    │ │
│  └────┬─────┘  └────┬─────┘  └─────┬──────┘ │
│       │              │              │        │
│  ┌────▼──────────────▼──────────────▼──────┐ │
│  │    Multi-Head Attention Extractor       │ │
│  │  (hook-based activation capture)        │ │
│  └────────────────┬────────────────────────┘ │
│                   │                          │
│  ┌────────────────▼────────────────────────┐ │
│  │    Visualization Renderer               │ │
│  │  (heatmap, bipartite, token-flow)       │ │
│  └─────────────────────────────────────────┘ │
└──────────────────────────────────────────────┘
        │
   ┌────▼──────────────┐
   │ Phase 28           │
   │ AttentionFilter    │
   │ ExecutiveController│
   └────────────────────┘
```

### Frozen Dataclasses

```python
@dataclass(frozen=True)
class AttentionMap:
    """Captured attention weights from a model layer."""
    layer_index: int
    head_index: int                          # -1 for averaged across heads
    weights: Tuple[Tuple[float, ...], ...]   # (seq_len, seq_len) attention matrix
    tokens: Tuple[str, ...] = ()             # token labels for visualization

@dataclass(frozen=True)
class RolloutResult:
    """Attention rollout aggregated across layers."""
    token_attributions: Tuple[float, ...]    # per-token attribution scores
    layer_contributions: Tuple[Tuple[float, ...], ...] # per-layer breakdown
    tokens: Tuple[str, ...] = ()
    method: str = "rollout"                  # "rollout" | "gradient_rollout"

@dataclass(frozen=True)
class RelevanceMap:
    """Layer-wise relevance propagation result."""
    input_relevance: Tuple[float, ...]       # per-input-dimension relevance
    layer_relevances: Tuple[Tuple[float, ...], ...] = ()  # per-layer
    conservation_error: float = 0.0          # relevance conservation check
    method: str = "lrp_epsilon"              # "lrp_epsilon" | "lrp_gamma" | "lrp_alpha_beta"

@dataclass(frozen=True)
class VisualizationConfig:
    """Configuration for attention visualization."""
    layers: Tuple[int, ...] = ()             # layers to visualize (empty = all)
    heads: Tuple[int, ...] = ()              # heads to visualize (empty = all)
    rollout_method: str = "standard"         # "standard" | "gradient_weighted"
    lrp_rule: str = "epsilon"               # "epsilon" | "gamma" | "alpha_beta"
    epsilon: float = 1e-6                    # LRP epsilon for numerical stability
    render_format: str = "heatmap"           # "heatmap" | "bipartite" | "sankey"
```

### Protocol Interface

```python
@runtime_checkable
class AttentionVisualizerProtocol(Protocol):
    """Protocol for attention visualization engines."""

    def extract_attention(
        self,
        model: Any,
        input_data: Any,
        layers: list[int] | None = None,
    ) -> list[AttentionMap]:
        """Extract attention maps from model layers."""
        ...

    def compute_rollout(
        self,
        attention_maps: list[AttentionMap],
        method: str = "standard",
    ) -> RolloutResult:
        """Compute attention rollout across layers."""
        ...

    def compute_lrp(
        self,
        model: Any,
        input_data: Any,
        target_class: int,
        rule: str = "epsilon",
    ) -> RelevanceMap:
        """Compute layer-wise relevance propagation."""
        ...

    def render(
        self,
        data: AttentionMap | RolloutResult | RelevanceMap,
        config: VisualizationConfig,
    ) -> bytes:
        """Render visualization as image bytes."""
        ...
```

### Acceptance Criteria

- [ ] Hook-based attention extraction for PyTorch transformer models
- [ ] Standard attention rollout with identity + attention matrix product
- [ ] Gradient-weighted attention rollout (Abnar & Zuidema 2020)
- [ ] GradCAM adapted for transformer classification heads
- [ ] LRP with ε-rule, γ-rule, and α-β rules
- [ ] BertViz-style multi-head bipartite graph visualization
- [ ] Relevance conservation verification (∑R_input ≈ f(x))
- [ ] Integration with Phase 28 AttentionFilter attention weights

### References

- Abnar & Zuidema (2020) — Quantifying Attention Flow in Transformers
- Selvaraju et al. (2017) — GradCAM: Visual Explanations from Deep Networks
- Bach et al. (2015) — On Pixel-Wise Explanations by Layer-Wise Relevance Propagation
- Vig (2019) — A Multiscale Visualization of Attention in the Transformer Model
- Chefer et al. (2021) — Transformer Interpretability Beyond Attention Visualization

**Comment by web3guru888** [comment]
## Implementation Notes — AttentionVisualizer Code Skeleton

```python
"""asi_build/xai/attention_visualizer.py — Phase 39.4"""
from __future__ import annotations

import numpy as np
from dataclasses import dataclass
from typing import Any

@dataclass(frozen=True)
class AttentionMap:
    layer_index: int
    head_index: int
    weights: tuple[tuple[float, ...], ...]
    tokens: tuple[str, ...] = ()

@dataclass(frozen=True)
class RolloutResult:
    token_attributions: tuple[float, ...]
    layer_contributions: tuple[tuple[float, ...], ...]
    tokens: tuple[str, ...] = ()
    method: str = "rollout"

@dataclass(frozen=True)
class RelevanceMap:
    input_relevance: tuple[float, ...]
    layer_relevances: tuple[tuple[float, ...], ...] = ()
    conservation_error: float = 0.0
    method: str = "lrp_epsilon"

@dataclass(frozen=True)
class VisualizationConfig:
    layers: tuple[int, ...] = ()
    heads: tuple[int, ...] = ()
    rollout_method: str = "standard"
    lrp_rule: str = "epsilon"
    epsilon: float = 1e-6
    render_format: str = "heatmap"


class AttentionVisualizer:
    """Attention visualization and LRP engine."""

    def extract_attention(self, model: Any, input_data: Any,
                          layers: list[int] | None = None) -> list[AttentionMap]:
        """Extract attention maps via forward hooks."""
        attention_maps = []
        captured = {}

        def hook_fn(layer_idx):
            def hook(module, input, output):
                # output.attentions shape: (batch, heads, seq, seq)
                captured[layer_idx] = output[1].detach().cpu().numpy()
            return hook

        # Register hooks
        handles = []
        target_layers = layers or list(range(model.config.num_hidden_layers))
        for idx in target_layers:
            h = model.encoder.layer[idx].attention.self.register_forward_hook(hook_fn(idx))
            handles.append(h)

        # Forward pass
        model(input_data)

        # Remove hooks
        for h in handles:
            h.remove()

        # Convert to AttentionMap objects
        for layer_idx, attn in captured.items():
            for head_idx in range(attn.shape[1]):
                w = attn[0, head_idx]  # (seq, seq)
                attention_maps.append(AttentionMap(
                    layer_index=layer_idx,
                    head_index=head_idx,
                    weights=tuple(tuple(row) for row in w.tolist()),
                ))

        return attention_maps

    def compute_rollout(self, attention_maps: list[AttentionMap],
                        method: str = "standard") -> RolloutResult:
        """Attention rollout: Abnar & Zuidema (2020)."""
        # Group by layer, average across heads
        layers: dict[int, list[np.ndarray]] = {}
        for am in attention_maps:
            if am.layer_index not in layers:
                layers[am.layer_index] = []
            layers[am.layer_index].append(np.array(am.weights))

        layer_avgs = {}
        for layer_idx, heads in sorted(layers.items()):
            avg = np.mean(heads, axis=0)
            # Add residual connection: 0.5 * attention + 0.5 * identity
            seq_len = avg.shape[0]
            avg = 0.5 * avg + 0.5 * np.eye(seq_len)
            # Normalize rows
            avg = avg / avg.sum(axis=-1, keepdims=True)
            layer_avgs[layer_idx] = avg

        # Multiply across layers
        rollout = None
        layer_contributions = []
        for layer_idx in sorted(layer_avgs.keys()):
            if rollout is None:
                rollout = layer_avgs[layer_idx]
            else:
                rollout = rollout @ layer_avgs[layer_idx]
            # CLS token (index 0) attribution to all input tokens
            layer_contributions.append(tuple(rollout[0].tolist()))

        token_attributions = tuple(rollout[0].tolist()) if rollout is not None else ()
        return RolloutResult(
            token_attributions=token_attributions,
            layer_contributions=tuple(layer_contributions),
            method=method,
        )

    def compute_lrp(self, model: Any, input_data: Any,
                    target_class: int, rule: str = "epsilon") -> RelevanceMap:
        """Layer-wise Relevance Propagation: Bach et al. (2015)."""
        eps = 1e-6 if rule == "epsilon" else 0.0
        gamma = 0.25 if rule == "gamma" else 0.0

        # Forward pass to collect activations
        activations = self._forward_collect(model, input_data)
        # Initialize relevance at output
        R = np.zeros(activations[-1].shape)
        R[target_class] = activations[-1][target_class]
        output_relevance = float(R.sum())

        layer_relevances = [tuple(R.tolist())]

        # Backward propagation of relevance
        for l in range(len(activations) - 2, -1, -1):
            a = activations[l]
            W = self._get_weights(model, l)
            z = np.outer(a, np.ones(R.shape[0])) * W  # a_j * w_jk
            if rule == "gamma":
                z_pos = np.maximum(z, 0)
                z = z + gamma * z_pos
            s = z.sum(axis=0) + eps * np.sign(z.sum(axis=0))
            R_new = (z / s[np.newaxis, :]) @ R
            R = R_new
            layer_relevances.append(tuple(R.tolist()))

        # Conservation check
        conservation_error = abs(float(R.sum()) - output_relevance) / max(abs(output_relevance), 1e-10)

        return RelevanceMap(
            input_relevance=tuple(R.tolist()),
            layer_relevances=tuple(reversed(layer_relevances)),
            conservation_error=conservation_error,
            method=f"lrp_{rule}",
        )

    def render(self, data: AttentionMap | RolloutResult | RelevanceMap,
               config: VisualizationConfig) -> bytes:
        """Render visualization as image bytes (placeholder for matplotlib)."""
        import io
        # In production: use matplotlib/plotly to generate actual visualizations
        buf = io.BytesIO()
        # ... rendering logic ...
        return buf.getvalue()

    def _forward_collect(self, model, input_data) -> list[np.ndarray]:
        """Collect layer activations during forward pass (placeholder)."""
        return model.forward_with_activations(input_data)

    def _get_weights(self, model, layer_index: int) -> np.ndarray:
        """Get weight matrix for a layer (placeholder)."""
        return model.get_layer_weights(layer_index)
```

### mypy type-check table

| Symbol | Type | Status |
|--------|------|--------|
| `AttentionVisualizer.extract_attention` | `(Any, Any, list[int]?) -> list[AttentionMap]` | ✅ |
| `AttentionVisualizer.compute_rollout` | `(list[AttentionMap], str) -> RolloutResult` | ✅ |
| `AttentionVisualizer.compute_lrp` | `(Any, Any, int, str) -> RelevanceMap` | ✅ |
| `AttentionVisualizer.render` | `(AttentionMap|Rollout|Relevance, VizConfig) -> bytes` | ✅ |
